src/text_shaper/directwrite_locator.cpp

```diff
@@ -0,0 +1,227 @@
+#include <text_shaper/font.h>
+
+#include <crispy/debuglog.h>
+
+#include <range/v3/view/iota.hpp>
+
+#include <string_view>
+
+ // {{{ TODO: replace with libunicode
+#include <codecvt>
+#include <locale>
+// }}}
+
+#include <wrl/client.h>
+#include <dwrite.h>
+#include <dwrite_3.h>
+
+using Microsoft::WRL::ComPtr;
+
+using std::nullopt;
+using std::optional;
+using std::string;
+using std::string_view;
+using std::unique_ptr;
+using std::vector;
+
+using namespace std::string_view_literals;
+
+namespace text
+{
+    namespace // {{{ support
+    {
+        font_weight dwFontWeight(int _weight)
+        {
+            switch (_weight)
+            {
+            case DWRITE_FONT_WEIGHT_THIN: return font_weight::thin;
+            case DWRITE_FONT_WEIGHT_EXTRA_LIGHT: return font_weight::extra_light;
+            case DWRITE_FONT_WEIGHT_LIGHT: return font_weight::light;
+            case DWRITE_FONT_WEIGHT_SEMI_LIGHT: return font_weight::demilight;
+            case DWRITE_FONT_WEIGHT_REGULAR: return font_weight::normal;
+                // XXX What about font_weight::book (which does exist via fontconfig)?
+            case DWRITE_FONT_WEIGHT_MEDIUM: return font_weight::medium;
+            case DWRITE_FONT_WEIGHT_DEMI_BOLD: return font_weight::demibold;
+            case DWRITE_FONT_WEIGHT_BOLD: return font_weight::bold;
+            case DWRITE_FONT_WEIGHT_EXTRA_BOLD: return font_weight::extra_bold;
+            case DWRITE_FONT_WEIGHT_BLACK: return font_weight::black;
+            case DWRITE_FONT_WEIGHT_EXTRA_BLACK: return font_weight::extra_black;
+            default: // TODO: the others
+                break;
+            }
+            return font_weight::normal; // TODO: rename normal to regular
+        }
+
+        font_slant dwFontSlant(int _style)
+        {
+            switch (_style)
+            {
+            case DWRITE_FONT_STYLE_NORMAL: return font_slant::normal;
+            case DWRITE_FONT_STYLE_ITALIC: return font_slant::italic;
+            case DWRITE_FONT_STYLE_OBLIQUE: return font_slant::oblique;
+            }
+            return font_slant::normal;
+        }
+
+        std::wstring get_font_path(IDWriteFontFace* fontFace)
+        {
+            ComPtr<IDWriteFontFile> fontFile;
+            UINT32 numberOfFiles = 0;
+            fontFace->GetFiles(&numberOfFiles, nullptr);
+            fontFace->GetFiles(&numberOfFiles, &fontFile);
+
+            ComPtr<IDWriteFontFileLoader> loader;
+            fontFile->GetLoader(&loader);
+            const void* key;
+            UINT32 keySize;
+            fontFile->GetReferenceKey(&key, &keySize);
+            ComPtr<IDWriteLocalFontFileLoader> localLoader;
+            loader.As(&localLoader);
+
+            UINT32 pathLen;
+            localLoader->GetFilePathLengthFromKey(key, keySize, &pathLen);
+            std::wstring path;
+            path.resize(pathLen);
+            localLoader->GetFilePathFromKey(key, keySize, path.data(), pathLen + 1);
```

Comment:
This returns the absolute path of the font. Same as font config locator.

Reply:
very nice!